# Basics - Data Analysis

#### Agenda

1. Define Business Problem
2. Collect/Import and Prepare Data
    - Data Collection
    - Data Cleaning
    - Data Transformation

3. Explore and Analyze
    - Exploratory Data Analysis (EDA)

4. Interpret Results and Communicate Findings
    - Visualization
    - Storytelling
    - Reporting

### 1. Define Business Problem

Here are several problem statements for HR datasets, categorized by common business challenges:

- Employee attrition and retention
    - Prediction of employee turnover
    - Root causes of attrition
    - Retention strategies for top talent

- Performance and productivity
    - Impact of training programs
    - Factors affecting employee performance
    - Predicting high-performers

- Workforce and skills gap analysis
    - Skills gap identification
    - Optimizing staffing levels
    - Recruitment channel effectiveness

- Employee well-being and engagement
    - Absenteeism and burnout
    - Improving employee engagement
    - DEI equity analysis

### 2. Collect/Import and Prepare Data

Here, we have a csv file having Human Resource data. This data has 1400+ records with 38 columns.

Understanding Dataset Columns

- Employee demographics
    - EmpID: Unique identifier for each employee.
    - Age: The employee's age in years.
    - AgeGroup: The employee's age, categorized into groups.
    - Gender: The employee's gender.
    - MaritalStatus: The employee's marital status.

- Organizational and role details
    - Department: The department where the employee works.
    - JobRole: The specific job role of the employee.
    - JobLevel: The employee's job level.
    - TotalWorkingYears: The total number of years the employee has worked professionally.
    - YearsAtCompany: The total number of years the employee has been with the current company.
    - YearsInCurrentRole: The number of years the employee has been in their current role.
    - YearsWithCurrManager: The number of years the employee has worked under their current manager.
    - YearsSinceLastPromotion: The number of years since the employee's last promotion.
    -   NumCompaniesWorked: The number of companies the employee has worked for prior to this one.
    - BusinessTravel: How frequently the employee travels for work.

- Compensation and performance
    - MonthlyIncome: The employee's monthly income.
    - SalarySlab: The monthly income categorized into salary brackets.
    - DailyRate: The employee's daily rate of pay.
    - HourlyRate: The employee's hourly rate of pay.
    - MonthlyRate: The employee's monthly rate of pay.
    - PercentSalaryHike: The percentage increase in the employee's most recent salary hike.
    - PerformanceRating: The employee's most recent performance review rating.
    - StockOptionLevel: The level of stock options the employee has been granted.
    - OverTime: Indicates if the employee works overtime (Yes/No). 

- Education and training
    - Education: The employee's highest level of education.
    - EducationField: The field of study for the employee's education.
    - TrainingTimesLastYear: The number of training sessions the employee attended in the previous year.

- Satisfaction and engagement
    - JobSatisfaction: The employee's satisfaction level with their job.
    - EnvironmentSatisfaction: The employee's satisfaction level with their work environment.
    - RelationshipSatisfaction: The employee's satisfaction level with their work relationships.
    - JobInvolvement: The employee's level of involvement in their job.
    - WorkLifeBalance: The employee's satisfaction with their work-life balance.

- Operational and attrition data
    - Attrition: Whether the employee has left the company (Yes/No).
    - DistanceFromHome: The distance in miles from the employee's home to the workplace.
    - EmployeeNumber: An additional, unique identifier for the employee.
    - EmployeeCount: The total count of employees (typically a constant value of 1 for each row).
    - Over18: Indicates if the employee's age is over 18 (typically a constant).
    - StandardHours: Standard working hours (likely a constant value).

In [1]:
# import required libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [22]:
# import data
# Data directory 📁 data » 📁 csv » 📁 <DATA-DOMAIN-FOLDER>

data = pd.read_csv("./data/csv/hr/HR_Analytics.csv")

In [23]:
# get basic information about the dataset

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1480 entries, 0 to 1479
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   EmpID                     1480 non-null   object 
 1   Age                       1480 non-null   int64  
 2   AgeGroup                  1480 non-null   object 
 3   Attrition                 1480 non-null   object 
 4   BusinessTravel            1480 non-null   object 
 5   DailyRate                 1480 non-null   int64  
 6   Department                1480 non-null   object 
 7   DistanceFromHome          1480 non-null   int64  
 8   Education                 1480 non-null   int64  
 9   EducationField            1480 non-null   object 
 10  EmployeeCount             1480 non-null   int64  
 11  EmployeeNumber            1480 non-null   int64  
 12  EnvironmentSatisfaction   1480 non-null   int64  
 13  Gender                    1480 non-null   object 
 14  HourlyRa

#### Data Cleaning

In [24]:
# check duplicate rows

data.duplicated().sum()

np.int64(7)

In [25]:
# check similarity between columns

data.nunique()

EmpID                       1470
Age                           43
AgeGroup                       5
Attrition                      2
BusinessTravel                 4
DailyRate                    886
Department                     3
DistanceFromHome              29
Education                      5
EducationField                 6
EmployeeCount                  1
EmployeeNumber              1470
EnvironmentSatisfaction        4
Gender                         2
HourlyRate                    71
JobInvolvement                 4
JobLevel                       5
JobRole                        9
JobSatisfaction                4
MaritalStatus                  3
MonthlyIncome               1349
SalarySlab                     4
MonthlyRate                 1427
NumCompaniesWorked            10
Over18                         1
OverTime                       2
PercentSalaryHike             15
PerformanceRating              2
RelationshipSatisfaction       4
StandardHours                  1
StockOptio

In [26]:
# lets set data into a DataFrame

df = pd.DataFrame(data)

In [27]:
df

EmpID  Age AgeGroup Attrition     BusinessTravel  DailyRate  \
0      RM297   18    18-25       Yes      Travel_Rarely        230   
1      RM302   18    18-25        No      Travel_Rarely        812   
2      RM458   18    18-25       Yes  Travel_Frequently       1306   
3      RM728   18    18-25        No         Non-Travel        287   
4      RM829   18    18-25       Yes         Non-Travel        247   
...      ...  ...      ...       ...                ...        ...   
1475   RM412   60      55+        No      Travel_Rarely        422   
1476   RM428   60      55+        No  Travel_Frequently       1499   
1477   RM537   60      55+        No      Travel_Rarely       1179   
1478   RM880   60      55+        No      Travel_Rarely        696   
1479  RM1210   60      55+        No      Travel_Rarely        370   

                  Department  DistanceFromHome  Education EducationField  ...  \
0     Research & Development                 3          3  Life Sciences  ...   
1                      Sales                10          3        Medical  ...   
2                      Sales                 5          3      Marketing  ...   
3     Research & Development                 5          2  Life Sciences  ...   
4     Research & Development                 8          1        Medical  ...   
...                      ...               ...        ...            ...  ...   
1475  Research & Development                 7          3  Life Sciences  ...   
1476                   Sales                28          3      Marketing  ...   
1477                   Sales                16          4      Marketing  ...   
1478                   Sales                 7          4      Marketing  ...   
1479  Research & Development                 1          4        Medical  ...   

      RelationshipSatisfaction  StandardHours  StockOptionLevel  \
0                            3             80                 0   
1                            1             80                 0   
2                            4             80                 0   
3                            4             80                 0   
4                            4             80                 0   
...                        ...            ...               ...   
1475                         4             80                 0   
1476                         4             80                 0   
1477                         4             80                 0   
1478                         2             80                 1   
1479                         3             80                 1   

     TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
0                    0                      2                3   
1                    0                      2                3   
2                    0                      3                3   
3                    0                      2                3   
4                    0                      0                3   
...                ...                    ...              ...   
1475                33                      5                1   
1476                22                      5                4   
1477                10                      1                3   
1478                12                      3                3   
1479                19                      2                4   

      YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
0                  0                  0                        0   
1                  0                  0                        0   
2                  0                  0                        0   
3                  0                  0                        0   
4                  0                  0                        0   
...              ...                ...                      ...   
1475              29                  8                       11   
1476              18                 13            

In [28]:
# lets get EmpID and EmployeeNumber columns as they are unique identifiers and not useful for analysis

df[['EmpID', 'EmployeeNumber']]

EmpID  EmployeeNumber
0      RM297             405
1      RM302             411
2      RM458             614
3      RM728            1012
4      RM829            1156
...      ...             ...
1475   RM412             549
1476   RM428             573
1477   RM537             732
1478   RM880            1233
1479  RM1210            1697

[1480 rows x 2 columns]

In [29]:
# check duplicate subset of EmpID and EmployeeNumber

df.duplicated(subset=['EmpID', 'EmployeeNumber']).sum()

np.int64(10)

In [ ]:
# check duplicate EmpID

df['EmpID'].duplicated().sum()

np.int64(10)

In [31]:
# check duplicate EmployeeNumber

df['EmployeeNumber'].duplicated().sum()

np.int64(10)

In [ ]:
# lets inspect the duplicate EmpID

df[df["EmpID"]=="RM1463"][["Age", "Department",'JobLevel', 'JobRole','TotalWorkingYears','YearsAtCompany']]

Age Department  JobLevel          JobRole  TotalWorkingYears  \
952   39      Sales         4  Sales Executive                 21   
954   39      Sales         4  Sales Executive                 21   

     YearsAtCompany  
952              20  
954              20

In [36]:
# here we can write a loop to check all duplicate EmpID but for now we will just check the other one

emp_ids = df[df['EmpID'].duplicated()]['EmpID'].unique()
for emp_id in emp_ids:
    print(f"Details for {emp_id}:")
    display(df[df['EmpID'] == emp_id])

Details for RM1465:


EmpID  Age AgeGroup Attrition BusinessTravel  DailyRate Department  \
161  RM1465   26    26-35        No  Travel_Rarely       1167      Sales   
162  RM1465   26    26-35        No  Travel_Rarely       1167      Sales   

     DistanceFromHome  Education EducationField  ...  \
161                 5          3          Other  ...   
162                 5          3          Other  ...   

     RelationshipSatisfaction  StandardHours  StockOptionLevel  \
161                         4             80                 0   
162                         4             80                 0   

    TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
161                 5                      2                3               4   
162                 5                      2                3               4   

    YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
161                  2                        0                  0.0  
162                  2                        0                  5.0  

[2 rows x 38 columns]

Details for RM1468:


EmpID  Age AgeGroup Attrition BusinessTravel  DailyRate  \
210  RM1468   27    26-35        No  Travel_Rarely        155   
211  RM1468   27    26-35        No  Travel_Rarely        155   

                 Department  DistanceFromHome  Education EducationField  ...  \
210  Research & Development                 4          3  Life Sciences  ...   
211  Research & Development                 4          3  Life Sciences  ...   

     RelationshipSatisfaction  StandardHours  StockOptionLevel  \
210                         2             80                 1   
211                         2             80                 1   

    TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
210                 6                      0                3               6   
211                 6                      0                3               6   

    YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
210                  2                        0                  3.0  
211                  2                        0                  3.0  

[2 rows x 38 columns]

Details for RM1461:


EmpID  Age AgeGroup Attrition BusinessTravel  DailyRate  \
327  RM1461   29    26-35        No  Travel_Rarely        468   
328  RM1461   29    26-35        No  Travel_Rarely        468   

                 Department  DistanceFromHome  Education EducationField  ...  \
327  Research & Development                28          4        Medical  ...   
328  Research & Development                28          4        Medical  ...   

     RelationshipSatisfaction  StandardHours  StockOptionLevel  \
327                         2             80                 0   
328                         2             80                 0   

    TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
327                 5                      3                1               5   
328                 5                      3                1               5   

    YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
327                  4                        0                  4.0  
328                  4                        0                  4.0  

[2 rows x 38 columns]

Details for RM1464:


EmpID  Age AgeGroup Attrition BusinessTravel  DailyRate  \
457  RM1464   31    26-35        No     Non-Travel        325   
458  RM1464   31    26-35        No     Non-Travel        325   

                 Department  DistanceFromHome  Education EducationField  ...  \
457  Research & Development                 5          3        Medical  ...   
458  Research & Development                 5          3        Medical  ...   

     RelationshipSatisfaction  StandardHours  StockOptionLevel  \
457                         2             80                 0   
458                         2             80                 0   

    TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
457                10                      2                3               9   
458                10                      2                3               9   

    YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
457                  4                        1                  7.0  
458                  4                        1                  7.0  

[2 rows x 38 columns]

Details for RM1470:


EmpID  Age AgeGroup Attrition BusinessTravel  DailyRate  \
654  RM1470   34    26-35        No   TravelRarely        628   
655  RM1470   34    26-35        No   TravelRarely        628   

                 Department  DistanceFromHome  Education EducationField  ...  \
654  Research & Development                 8          3        Medical  ...   
655  Research & Development                 8          3        Medical  ...   

     RelationshipSatisfaction  StandardHours  StockOptionLevel  \
654                         1             80                 0   
655                         1             80                 0   

    TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
654                 6                      3                4               4   
655                 6                      3                4               4   

    YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
654                  3                        1                  2.0  
655                  3                        1                  2.0  

[2 rows x 38 columns]

Details for RM1466:


EmpID  Age AgeGroup Attrition     BusinessTravel  DailyRate  \
802  RM1466   36    36-45        No  Travel_Frequently        884   
803  RM1466   36    36-45        No  Travel_Frequently        884   

                 Department  DistanceFromHome  Education EducationField  ...  \
802  Research & Development                23          2        Medical  ...   
803  Research & Development                23          2        Medical  ...   

     RelationshipSatisfaction  StandardHours  StockOptionLevel  \
802                         3             80                 1   
803                         3             80                 1   

    TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
802                17                      3                3               5   
803                17                      3                3               5   

    YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
802                  2                        0                  3.0  
803                  2                        0                  2.0  

[2 rows x 38 columns]

Details for RM1463:


EmpID  Age AgeGroup Attrition BusinessTravel  DailyRate Department  \
952  RM1463   39    36-45        No  Travel_Rarely        722      Sales   
954  RM1463   39    36-45        No  Travel_Rarely        722      Sales   

     DistanceFromHome  Education EducationField  ...  \
952                24          1      Marketing  ...   
954                24          1      Marketing  ...   

     RelationshipSatisfaction  StandardHours  StockOptionLevel  \
952                         1             80                 1   
954                         1             80                 1   

    TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
952                21                      2                2              20   
954                21                      2                2              20   

    YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
952                  9                        9                  6.0  
954                  9                        9                  6.0  

[2 rows x 38 columns]

Details for RM1467:


EmpID  Age AgeGroup Attrition BusinessTravel  DailyRate  \
953  RM1467   39    36-45        No  Travel_Rarely        613   
955  RM1467   39    36-45        No  Travel_Rarely        613   

                 Department  DistanceFromHome  Education EducationField  ...  \
953  Research & Development                 6          1        Medical  ...   
955  Research & Development                 6          1        Medical  ...   

     RelationshipSatisfaction  StandardHours  StockOptionLevel  \
953                         1             80                 1   
955                         1             80                 1   

    TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
953                 9                      5                3               7   
955                 9                      5                3               7   

    YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
953                  7                        1                  7.0  
955                  7                        1                  1.0  

[2 rows x 38 columns]

Details for RM1469:


EmpID  Age AgeGroup Attrition     BusinessTravel  DailyRate Department  \
1304  RM1469   49    46-55        No  Travel_Frequently       1023      Sales   
1305  RM1469   49    46-55        No  Travel_Frequently       1023      Sales   

      DistanceFromHome  Education EducationField  ...  \
1304                 2          3        Medical  ...   
1305                 2          3        Medical  ...   

      RelationshipSatisfaction  StandardHours  StockOptionLevel  \
1304                         4             80                 0   
1305                         4             80                 0   

     TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
1304                17                      3                2   
1305                17                      3                2   

      YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
1304               9                  6                        0   
1305               9                  6                        0   

     YearsWithCurrManager  
1304                  8.0  
1305                  8.0  

[2 rows x 38 columns]

Details for RM1462:


EmpID  Age AgeGroup Attrition BusinessTravel  DailyRate Department  \
1335  RM1462   50    46-55       Yes  Travel_Rarely        410      Sales   
1336  RM1462   50    46-55       Yes  Travel_Rarely        410      Sales   

      DistanceFromHome  Education EducationField  ...  \
1335                28          3      Marketing  ...   
1336                28          3      Marketing  ...   

      RelationshipSatisfaction  StandardHours  StockOptionLevel  \
1335                         2             80                 1   
1336                         2             80                 1   

     TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
1335                20                      3                3   
1336                20                      3                3   

      YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
1335               3                  2                        2   
1336               3                  2                        2   

     YearsWithCurrManager  
1335                  0.0  
1336                  0.0  

[2 rows x 38 columns]

We have to fix this 

In [39]:
# Now you can see that the duplicate EmpIDs have different details in other columns.
# This indicates that these EmpIDs are not unique identifiers for employees, as they are associated with different records.

# We have to fix this by removing duplicate EmpIDs by keeping the first occurrence

df = df.drop_duplicates(subset=['EmpID'], keep='first')

In [ ]:
# check again duplicate EmpID

df['EmpID'].duplicated().sum()

np.int64(0)

In [42]:
# check duplicate entire rows again

df.duplicated().sum()

np.int64(0)

Here, our dataset has `0` duplicate records. We have successfully removed duplicate EmpIDs and there are no duplicate rows in the dataset.

Next, we have to check missing values.

In [43]:
# check missing values

df.isnull().sum()

EmpID                        0
Age                          0
AgeGroup                     0
Attrition                    0
BusinessTravel               0
DailyRate                    0
Department                   0
DistanceFromHome             0
Education                    0
EducationField               0
EmployeeCount                0
EmployeeNumber               0
EnvironmentSatisfaction      0
Gender                       0
HourlyRate                   0
JobInvolvement               0
JobLevel                     0
JobRole                      0
JobSatisfaction              0
MaritalStatus                0
MonthlyIncome                0
SalarySlab                   0
MonthlyRate                  0
NumCompaniesWorked           0
Over18                       0
OverTime                     0
PercentSalaryHike            0
PerformanceRating            0
RelationshipSatisfaction     0
StandardHours                0
StockOptionLevel             0
TotalWorkingYears            0
Training

We can see above that are `57` missing values in the dataset with respect to columns `YearsWithCurrManager`.

In [45]:
df['YearsWithCurrManager'].isnull().sum()

np.int64(57)

Lets check the possibilities wether we can manage those missing values in column `YearsWithCurrManager` statistically or not rather than removing that column.

In [ ]:
# check the differences between Filled and Unfilled records in column YearsWithCurrManager
df_filled = df['YearsWithCurrManager'].notnull().sum()
df_unfilled = df['YearsWithCurrManager'].isnull().sum()

total_records = len(df)

percentage_unfilled = (df_unfilled / total_records) * 100

print(f"Filled records: {df_filled}")
print(f"Unfilled records: {df_unfilled}")
print(f"Total records: {total_records}")
print(f"Percentage of unfilled records: {percentage_unfilled:.2f}%")

Filled records: 1413
Unfilled records: 57
Total records: 1470
Percentage of unfilled records: 3.88%


Total unfilled records are `57` out of `1470` which is approximately `3.88%` of the total records.

This is a small percentage, so we can consider removing these records without significantly impacting the dataset.

To fill the missing values in the `YearsWithCurrManager` column, we can use the `median` value of the filled records.

In [49]:
median_value = df['YearsWithCurrManager'].median()
df['YearsWithCurrManager'].fillna(median_value, inplace=True)

C:\Users\ikjan\AppData\Local\Temp\ipykernel_50184\4035732794.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['YearsWithCurrManager'].fillna(median_value, inplace=True)
C:\Users\ikjan\AppData\Local\Temp\ipykernel_50184\4035732794.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['YearsWithCurrManager'].fillna(median_value, inplace=T

In [50]:
# Verify that there are no more missing values in the 'YearsWithCurrManager' column
df['YearsWithCurrManager'].isnull().sum()

np.int64(0)

In [51]:
df['YearsWithCurrManager']

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
1475    10.0
1476    11.0
1477     2.0
1478     9.0
1479     0.0
Name: YearsWithCurrManager, Length: 1470, dtype: float64

Done! All missing values filled with `median`.

Now, get `stats` of the dataset.

In [52]:
df.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000  1470.000000         1470.0   
mean     36.923810   802.485714          9.192517     2.912925            1.0   
std       9.135373   403.509100          8.106864     1.024165            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   802.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  ...  RelationshipSatisfaction  StandardHours  \
count  1470.000000  ...               1470.000000         1470.0   
mean      2.063946  ...                  2.712245           80.0   
std       1.106940  ...                  1.081209            0.0   
min       1.000000  ...                  1.000000           80.0   
25%       1.000000  ...                  2.000000           80.0   
50%       2.000000  ...                  3.000000           80.0   
75%       3.000000  ...                  4.000000           80.0   
max       5.000000  ...                  4.000000           80.0   

       StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
count       1470.000000        1470.000000            1470.000000   
mean           0.793878          11.279592               2.799320   
std            0.852077           7.780782               1.289271   
min            0.000000           0.000000               0.000000   
25%            0.000000           6.000000               2.000000   
50%            1.000000          10.000000               3.000000   
75%            1.000000          15.000000               3.000000   
max            3.000000          40.000000               6.000000   

       WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
count      1470.000000     1470.000000         1470.000000   
mean          2.761224        7.008163            4.229252   
std           0.706476        6.126525            3.623137   
min           1.000000        0.000000            0.000000   
25%           2.000000        3.000000            2.000000   
50%           3.000000        5.000000            3.000000   
75%           3.000000        9.000000            7.000000   
max           4.000000       40.000000           18.000000   

       YearsSinceLastPromotion  YearsWithCurrManager  
count              1470.000000           1470.000000  
mean                  2.187755              4.076871  
std                   3.222430              3.498547  
min                   0.000000              0.000000  
25%                   0.000000              2.000000  
50%                   1.000000              3.000000  
75%                   3.000000              7.000000  
max                  15.000000             17.000000  

[8 rows x 26 columns]

In [53]:
df.describe().T

count          mean          std     min      25%  \
Age                       1470.0     36.923810     9.135373    18.0    30.00   
DailyRate                 1470.0    802.485714   403.509100   102.0   465.00   
DistanceFromHome          1470.0      9.192517     8.106864     1.0     2.00   
Education                 1470.0      2.912925     1.024165     1.0     2.00   
EmployeeCount             1470.0      1.000000     0.000000     1.0     1.00   
EmployeeNumber            1470.0   1024.865306   602.024335     1.0   491.25   
EnvironmentSatisfaction   1470.0      2.721769     1.093082     1.0     2.00   
HourlyRate                1470.0     65.891156    20.329428    30.0    48.00   
JobInvolvement            1470.0      2.729932     0.711561     1.0     2.00   
JobLevel                  1470.0      2.063946     1.106940     1.0     1.00   
JobSatisfaction           1470.0      2.728571     1.102846     1.0     2.00   
MonthlyIncome             1470.0   6502.931293  4707.956783  1009.0  2911.00   
MonthlyRate               1470.0  14313.103401  7117.786044  2094.0  8047.00   
NumCompaniesWorked        1470.0      2.693197     2.498009     0.0     1.00   
PercentSalaryHike         1470.0     15.209524     3.659938    11.0    12.00   
PerformanceRating         1470.0      3.153741     0.360824     3.0     3.00   
RelationshipSatisfaction  1470.0      2.712245     1.081209     1.0     2.00   
StandardHours             1470.0     80.000000     0.000000    80.0    80.00   
StockOptionLevel          1470.0      0.793878     0.852077     0.0     0.00   
TotalWorkingYears         1470.0     11.279592     7.780782     0.0     6.00   
TrainingTimesLastYear     1470.0      2.799320     1.289271     0.0     2.00   
WorkLifeBalance           1470.0      2.761224     0.706476     1.0     2.00   
YearsAtCompany            1470.0      7.008163     6.126525     0.0     3.00   
YearsInCurrentRole        1470.0      4.229252     3.623137     0.0     2.00   
YearsSinceLastPromotion   1470.0      2.187755     3.222430     0.0     0.00   
YearsWithCurrManager      1470.0      4.076871     3.498547     0.0     2.00   

                              50%       75%      max  
Age                          36.0     43.00     60.0  
DailyRate                   802.0   1157.00   1499.0  
DistanceFromHome              7.0     14.00     29.0  
Education                     3.0      4.00      5.0  
EmployeeCount                 1.0      1.00      1.0  
EmployeeNumber             1020.5   1555.75   2068.0  
EnvironmentSatisfaction       3.0      4.00      4.0  
HourlyRate                   66.0     83.75    100.0  
JobInvolvement                3.0      3.00      4.0  
JobLevel                      2.0      3.00      5.0  
JobSatisfaction               3.0      4.00      4.0  
MonthlyIncome              4919.0   8379.00  19999.0  
MonthlyRate               14235.5  20461.50  26999.0  
NumCompaniesWorked            2.0      4.00      9.0  
PercentSalaryHike            14.0     18.00     25.0  
PerformanceRating             3.0      3.00      4.0  
RelationshipSatisfaction      3.0      4.00      4.0  
StandardHours                80.0     80.00     80.0  
StockOptionLevel              1.0      1.00      3.0  
TotalWorkingYears            10.0     15.00     40.0  
TrainingTimesLastYear         3.0      3.00      6.0  
WorkLifeBalance               3.0      3.00      4.0  
YearsAtCompany                5.0      9.00     40.0  
YearsInCurrentRole            3.0      7.00     18.0  
YearsSinceLastPromotion       1.0      3.00     15.0  
YearsWithCurrManager          3.0      7.00     17.0

Till this point, we have done the basic `Data Cleaning` and preprocessing. Next, we can move on to `Exploratory Data Analysis (EDA)` and `Visualization` to gain insights from the dataset.

### 3. Explore and Analyze

To understand this step, click *[Data analytics workflow](https://github.com/imkjangid/Code2Insights#data-analytics-workflow)*.

#### Exploratory Data Analysis (EDA)

Step-1: Get demographic details and counts

In [55]:
# demographic details and counts with Age, AgeGroup, Gender and MaritalStatus by grouping them
demographic_counts = df.groupby(['AgeGroup', 'Gender', 'MaritalStatus']).size()
demographic_counts

AgeGroup  Gender  MaritalStatus
18-25     Female  Divorced           4
                  Married           11
                  Single            28
          Male    Divorced          11
                  Married           34
                  Single            35
26-35     Female  Divorced          49
                  Married          104
                  Single            85
          Male    Divorced          87
                  Married          164
                  Single           117
36-45     Female  Divorced          40
                  Married           89
                  Single            57
          Male    Divorced          75
                  Married          131
                  Single            76
46-55     Female  Divorced          21
                  Married           54
                  Single            25
          Male    Divorced          28
                  Married           63
                  Single            35
55+       Female  Divorced      

In [67]:
# demographic details and counts with Age, AgeGroup, Gender and MaritalStatus by subsetting them
demographic_counts_sub = df[['Age']].copy()
demographic_counts_sub['AgeGroup'] = df['AgeGroup']
demographic_counts_sub['Gender'] = df['Gender']
demographic_counts_sub['MaritalStatus'] = df['MaritalStatus']
demographic_counts_sub = demographic_counts_sub.groupby(['AgeGroup', 'Gender', 'MaritalStatus']).sum()
demographic_counts_sub = demographic_counts_sub.rename(columns={'Age': 'TotalCount'})
demographic_counts_sub

TotalCount
AgeGroup Gender MaritalStatus            
18-25    Female Divorced               97
                Married               258
                Single                586
         Male   Divorced              260
                Married               811
                Single                744
26-35    Female Divorced             1497
                Married              3220
                Single               2661
         Male   Divorced             2687
                Married              5112
                Single               3602
36-45    Female Divorced             1611
                Married              3530
                Single               2273
         Male   Divorced             2991
                Married              5267
                Single               3029
46-55    Female Divorced             1043
                Married              2691
                Single               1269
         Male   Divorced             1381
                Married              3197
                Single               1747
55+      Female Divorced              170
                Married               808
                Single                236
         Male   Divorced              522
                Married               517
                Single                461

In [ ]:
# count of Employees by Age and Gender
demographic_counts_sub = df[['Age', 'Gender']].copy()
demographic_counts_sub['TotalCount'] = 1
demographic_counts_sub = demographic_counts_sub.groupby(['Age', 'Gender']).sum()
demographic_counts_sub

TotalCount
Age Gender            
18  Female           4
    Male             4
19  Female           4
    Male             5
20  Female           6
...                ...
58  Male             8
59  Female           6
    Male             4
60  Female           2
    Male             3

[85 rows x 1 columns]